fix: enable the reporting of parsing related problems when running deno lint

HasanAlrimawi opened the pull request:
Supported reporting of syntax/parsing errors.

The commit is made to address issue #23437 

Code changes

Added check for ParsedSource.diagnostics() and when it's not empty, all parse diagnostic errors will be logged/displayed.

PS: The change was within `fn handle_lint_result()` to log any found parse or lint diagnostic errors, since this function is called after any lint is done.

bartlomieju reviewed (commented):
Thanks, could you add a test to `tests/specs/lint/` that exercises this change?

dsherret commented on cli/tools/lint/mod.rs:
```diff
@@ -525,7 +525,16 @@ fn handle_lint_result(
+    Ok((source, mut file_diagnostics)) => {
+      if !source.diagnostics().is_empty() {
+        for parse_diagnostic in source.diagnostics() {
+          log::info!("{}: {}", colors::red("error"), parse_diagnostic);
```
Maybe this should be `colors::yellow("warn")` since it's not a fatal error?

Also, perhaps we should just remove the "found x parsing related problems" message?

HasanAlrimawi replied:
Surely, suggested changes are done and pushed.
I'd also like to point out that I previously aimed to follow the same approach used for reporting/logging Lint diagnostic errors (which logs error in red color and prints the number of Lint diagnostic errors).

HasanAlrimawi commented:
Hello @bartlomieju ,
I added a test to exercise the change, but the PR check _(ci / lint debug linux-x86_64)_ failed because the test I added includes syntax error (unterminated string constant) in the script file. However, this syntax error was intentionally added to show that the change done can detect this syntax error.
Is there a way to overcome this from my or your side, or should I just remove the test?
![PR check failed](https://github.com/denoland/deno/assets/141642411/e1beba95-f0f1-478f-a740-e22afb9eae5f)

HasanAlrimawi commented:
Hello @bartlomieju ,
Just as a notification.
What do you think is best to do with the case [here](https://github.com/denoland/deno/pull/24332#issuecomment-2200038308)

bartlomieju commented:
@HasanAlrimawi you need to update `.dprint.json` file and add the file you added to the `"excludes"` config so that it's not being formatted.

dsherret commented on .dprint.json:
```diff
@@ -12,10 +12,12 @@
-      "command": "rustfmt --config imports_granularity=item",
-      "exts": ["rs"]
-    }]
+    "commands": [
+      {
+        "command": "rustfmt --config imports_granularity=item",
+        "exts": ["rs"]
+      }
+    ]
```
```suggestion
    "commands": [{
      "command": "rustfmt --config imports_granularity=item",
      "exts": ["rs"]
    }]
```

dsherret commented on cli/tools/lint/mod.rs:
```diff
@@ -646,7 +646,12 @@ fn handle_lint_result(
+    Ok((source, mut file_diagnostics)) => {
+      if !source.diagnostics().is_empty() {
+        for parse_diagnostic in source.diagnostics() {
+          log::info!("{}: {}", colors::yellow("warn"), parse_diagnostic);
```
```suggestion
          log::warn!("{}: {}", colors::yellow("warn"), parse_diagnostic);
```

dsherret reviewed (approved):
LGTM. Thanks!
